src/webnn_native/GraphBuilder.cpp

```diff
@@ -42,92 +43,106 @@
                                            size_t new_shape_count) {
         DAWN_VALIDATE_AND_INFER_TYPES(new op::Reshape(this, input, new_shape, new_shape_count));
     }
 
-    OperandBase* ModelBuilderBase::Softmax(OperandBase* input) {
+    OperandBase* GraphBuilderBase::Softmax(OperandBase* input) {
         DAWN_VALIDATE_AND_INFER_TYPES(new op::Unary(this, op::UnaryOpType::kSoftmax, input));
     }
 
-    OperandBase* ModelBuilderBase::Transpose(OperandBase* input, TransposeOptions const* options) {
+    OperandBase* GraphBuilderBase::Transpose(OperandBase* input, TransposeOptions const* options) {
         DAWN_VALIDATE_AND_INFER_TYPES(new op::Transpose(this, input, options));
     }
 
-    ModelBase* ModelBuilderBase::CreateModel(NamedOperandsBase const* namedOperands) {
-        Ref<ModelBase> model = AcquireRef(CreateModelImpl());
+    void GraphBuilderBase::Build(NamedOperandsBase const* namedOperands,
+                                 MLBuildCallback callback,
+                                 void* userdata) {
+        if (DAWN_UNLIKELY(this->IsError())) {
+            BUILD_ERROR_AND_CALLBACK("This Graph object is an error");
+        }
+
         std::vector<const OperandBase*> outputs;
         if (namedOperands->GetRecords().empty()) {
-            return ModelBase::MakeError(this);
+            BUILD_ERROR_AND_CALLBACK("The output named operands are empty.");
         }
         for (auto& namedOutput : namedOperands->GetRecords()) {
             outputs.push_back(namedOutput.second);
         }
         std::vector<const OperandBase*> sorted_operands = TopologicalSort(outputs);
+        Ref<GraphBase> graph = AcquireRef(GetContext()->CreateGraph());
         for (auto& op : sorted_operands) {
-            if (op->IsError() || GetContext()->ConsumedError(op->AddToModel(model.Get()))) {
-                return ModelBase::MakeError(this);
+            if (op->IsError() || GetContext()->ConsumedError(op->AddToGraph(graph.Get()))) {
+                BUILD_ERROR_AND_CALLBACK("Failed to add the operand when building graph.");
             }
         }
         for (auto& namedOutput : namedOperands->GetRecords()) {
             if (GetContext()->ConsumedError(
-                    model->AddOutput(namedOutput.first, namedOutput.second))) {
-                return ModelBase::MakeError(this);
+                    graph->AddOutput(namedOutput.first, namedOutput.second))) {
+                BUILD_ERROR_AND_CALLBACK("Failed to add output when building graph.");
             }
         }
-        if (GetContext()->ConsumedError(model->Finish())) {
-            return ModelBase::MakeError(this);
+        if (GetContext()->ConsumedError(graph->Finish())) {
+            BUILD_ERROR_AND_CALLBACK("Failed to finish building graph.");
         }
-        return model.Detach();
+        // Compile the graph, call back the graph to user if sucessful, otherwise delete it in async
+        // thread.
```

Comment:
Could you please point me the code that deletes the graph if compile fails?

Reply:
Add lamda function to delete failed graph in the call back.